pkg/apis/internal.admin.acorn.io/v1/resources.go

```diff
@@ -7,65 +7,97 @@ import (
-	Secrets    int `json:"secrets,omitempty"`
-	Images     int `json:"images,omitempty"`
-	Projects   int `json:"projects,omitempty"`
-
-	VolumeStorage resource.Quantity `json:"volumeStorage,omitempty"`
-	Memory        resource.Quantity `json:"memory,omitempty"`
-	CPU           resource.Quantity `json:"cpu,omitempty"`
+	Flags            Flags            `json:"flags,omitempty"`
+	Counts           Counts           `json:"counts,omitempty"`
+	PersistentCounts PersistentCounts `json:"persistentCounts,omitempty"`
+	Quantities       Quantities       `json:"quantities,omitempty"`
+}
+
+// NewResources initializes and returns a new Resources instance with all maps initialized.
+func NewResources() Resources {
+	return Resources{
+		Flags:            make(Flags),
+		Counts:           make(Counts),
+		PersistentCounts: make(PersistentCounts),
+		Quantities:       make(Quantities),
+	}
 }
 
 // Add will add the resources of another Resources struct into the current one.
 func (current *Resources) Add(incoming Resources) {
-	current.Apps += incoming.Apps
-	current.Containers += incoming.Containers
-	current.Jobs += incoming.Jobs
-	current.Volumes += incoming.Volumes
-	current.Secrets += incoming.Secrets
-	current.Images += incoming.Images
-	current.Projects += incoming.Projects
-
-	current.VolumeStorage.Add(incoming.VolumeStorage)
-	current.Memory.Add(incoming.Memory)
-	current.CPU.Add(incoming.CPU)
+	// Add the incoming resources to the current ones
+	for flag, value := range incoming.Flags {
+		current.Flags[flag] = value
+	}
+	for count, value := range incoming.Counts {
+		current.Counts[count] += value
+	}
+	for quantity, value := range incoming.Quantities {
+		q := current.Quantities[quantity]
+		q.Add(value)
+	}
+	for persistentCount, value := range incoming.PersistentCounts {
+		current.PersistentCounts[persistentCount] += value
+	}
 }
 
 // Remove will remove the resources of another Resources struct from the current one.
 func (current *Resources) Remove(incoming Resources, all bool) {
-	// Do not allow resources to go below 0
-	nonNegativeSubtract := func(currentVal, incomingVal int) int {
-		difference := currentVal - incomingVal
-		if difference < 0 {
-			difference = 0
+	nonZeroSubtract := func(a, b int) int {
+		if a > b {
+			return a - b
 		}
-		return difference
+		return 0
 	}
 
-	current.Apps = nonNegativeSubtract(current.Apps, incoming.Apps)
-	current.Containers = nonNegativeSubtract(current.Containers, incoming.Containers)
-	current.Jobs = nonNegativeSubtract(current.Jobs, incoming.Jobs)
-	current.Volumes = nonNegativeSubtract(current.Volumes, incoming.Volumes)
-	current.Images = nonNegativeSubtract(current.Images, incoming.Images)
-	current.Projects = nonNegativeSubtract(current.Projects, incoming.Projects)
+	for c, value := range incoming.Counts {
+		current.Counts[c] = nonZeroSubtract(current.Counts[c], value)
+	}
+	for quantity, value := range incoming.Quantities {
+		q := current.Quantities[quantity]
+		q.Sub(value)
```

Comment:
Same question here.

Reply:
Answer above ^ Thanks for bringing this up.